### PR TITLE
parseFeatureCollection sorts features by globalId

### PR DESCRIPTION
**Pull request by felixpalmer**
Just taking the extra changes from https://github.com/visgl/loaders.gl/pull/1640 and copying across @ibgreen comments

### CODE REVIEW

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 1]
```diff
@@ -172,18 +173,42 @@ function parseFeatureCollection(dataArray): Feature[] {
...
+
   return features;
 }
 
+function _checkIdRange(featuresById, expectedCount) {
+  if (featuresById.size === 0) {
+    return;
+  }
+
+  const minId = Math.min(...featuresById.keys());
```
Comment (felixpalmer) [inline modules/gis/src/lib/binary-to-geojson.ts:205]: Rest arguments crash for long arrays. We have had bugs around this in the past.

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 2]
```diff
@@ -156,7 +156,8 @@ function deduceReturnType(dataArray): 'FeatureCollection' | 'Geometry' {
 
 /** Parse input binary data and return an array of GeoJSON Features */
 function parseFeatureCollection(dataArray): Feature[] {
-  const features: Feature[] = [];
+  const featuresById: Map<number, Feature> = new Map<number, Feature>();
```
Comment (felixpalmer) [inline modules/gis/src/lib/binary-to-geojson.ts:159]: If we know all the ids are covered, all we have to do is to create an array of the right length.
Suggested change:
  const featuresById: Map<number, Feature> = new Map<number, Feature>();
  const featuresById = new Array<Feature>(globalFeatureIds.length).fill(null);

This map business would seem to affect perf and memory.

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 3]
```diff
@@ -172,18 +173,42 @@ function parseFeatureCollection(dataArray): Feature[] {
...
+  _checkIdRange(featuresById, count);
+  const features: Feature[] = new Array<Feature>(featuresById.size);
+  for (const [featureId, feature] of featuresById) {
+    features[featureId] = feature;
   }
+
   return features;
 }
 
+function _checkIdRange(featuresById, expectedCount) {
```
Comment (felixpalmer) [inline modules/gis/src/lib/binary-to-geojson.ts:200]: Nit: we don't use underscore on private function. The absence of export already makes this clear.

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 2 above)
Comment (felixpalmer) [reply modules/gis/src/lib/binary-to-geojson.ts:159]: The issue is that we don't know the length until we've scanned through all the features, that's why I collect the features mapped to their original id and then create the array.

Comment (felixpalmer) [reply modules/gis/src/lib/binary-to-geojson.ts:159]: Let's resolve this first, and then I can look at your other comments if they are still relevant

**Comment by ibgreen** [comment]
Close due to inactivity

**Pull request closed without merging**